src/echo-common/src/smart-cache.ts

```diff
@@ -0,0 +1,186 @@
+import { BehaviorSubject, filter, groupBy, map, mergeMap, Observable, of, Subject, take } from 'rxjs'
+import { concatMap, delay, mergeAll, throttleTime } from 'rxjs/operators'
+import { filterNullish } from 'ts-ratchet'
+import { EchoDirService } from './echo-dir-service'
+
+export type SmartCacheLoadMode = "Default" | "CacheOnly" | "NoCache"
+
+export type SmartCacheLoadConfig = {
+  key: string,
+  mode?: SmartCacheLoadMode
+}
+
+export type SmartCacheBaseEvent = {
+  key: string,
+  timestampMs: number,
+}
+
+export type SmartCacheResultEvent<T> = SmartCacheBaseEvent & {
+  type: "result",
+  source: "cache-local" | "cache-memory" | "live",
+  result: T | null | undefined,
+}
+
+export type SmartCacheLoadingEvent = SmartCacheBaseEvent & {
+  type: "loading",
+}
+
+export type SmartCacheErrorEvent = SmartCacheBaseEvent & {
+  type: "error",
+  error: any
+}
+
+export type SmartCacheRateLimitEvent = SmartCacheBaseEvent & {
+  type: "rate-limit",
+  limitExpiresMs: number
+}
+
+export type SmartCacheStatusEvent = SmartCacheLoadingEvent | SmartCacheErrorEvent | SmartCacheRateLimitEvent
+
+export type SmartCacheEvent<T> = SmartCacheStatusEvent | SmartCacheResultEvent<T>
+
+export type SmartCacheStore<T> = {
+  lastResultEvent: SmartCacheResultEvent<T> | undefined,
+  lastStatusEvent: SmartCacheStatusEvent | undefined,
+}
+
+export type SmartCacheJob = {
+  config: SmartCacheLoadConfig
+}
+
+export class SmartCache<T> {
+  private jobQueue$ = new Subject<SmartCacheJob>()
+  private events$ = new Subject<SmartCacheEvent<T>>()
```

Comment:
I think, if I want to add user-actions or actions which are triggered outside of react, I would attach `pipe` and subscribe here? But this subjects are `private`, so how do I do that?
I think it would make sense to add a basic `pipe` without to subscribe?